crates/datagen/src/lib.rs

```diff
@@ -337,194 +351,382 @@ impl InputGenerator {
+                        rows: consumed,
+                        todo: metadata_todo,
+                    }
+                    .into();
+                    consumer.extended(num_records, rmpv::ext::to_value(metadata).unwrap());
+                }
+                Some(InputReaderCommand::Disconnect) => break,
+                None => (),
+            }
+
+            while let Ok(completion) = completion_receiver.try_recv() {
+                let batch = &completion.batch;
+                in_flight[batch.plan_idx].remove_range(batch.rows.start..=batch.rows.end - 1);
+                consumer.buffered(completion.buffer.len(), 0);
+                completed.push_back(completion);
+            }
 
-    fn unpark(&self) {
-        self.notifier.notify_waiters()
+            if running && !work_sender.is_full() {
+                if let Some(work) = assign_work(&mut unassigned, &mut in_flight, &config, true) {
+                    let _ = work_sender.send_blocking(work);
+                } else if in_flight.iter().all(|set| set.is_empty()) && !eoi {
+                    eoi = true;
+                    running = false;
+                    consumer.eoi();
+                }
+            } else {
+                std::thread::park();
+            }
+        }
+
+        Ok(())
     }
 
-    #[allow(clippy::too_many_arguments, clippy::type_complexity)]
+    #[allow(clippy::too_many_arguments)]
     async fn worker_thread(
-        worker_idx: usize,
+        work_receiver: AsyncReceiver<Work>,
+        completion_sender: TokioSender<Completion>,
         config: DatagenInputConfig,
-        shared_state: Arc<
-            Vec<(
-                AtomicUsize,
-                RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>,
-            )>,
-        >,
         schema: Relation,
         consumer: Box<dyn InputConsumer>,
         mut parser: Box<dyn Parser>,
-        queue: Arc<InputQueue>,
-        notifier: Arc<Notify>,
-        status: Arc<Atomic<PipelineState>>,
-        generated: Arc<AtomicUsize>,
+        rate_limiters: Arc<Vec<RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>>>,
+        datagen_thread: Thread,
     ) {
         let mut buffer = Vec::new();
+        let mut buffer_start = 0;
         static START_ARR: &[u8; 1] = b"[";
         static END_ARR: &[u8; 1] = b"]";
         static REC_DELIM: &[u8; 1] = b",";
 
-        for (plan, (progress, rate_limiter)) in config.plan.into_iter().zip(shared_state.iter()) {
-            // Inserting in batches improves performance by around ~80k records/s for
-            // me, so we always do it rather than giving the user the option.
-            // If we have a low rate, it's necessary to adjust the batch-size down otherwise no inserts might be visible
-            // for a long time (until a batch is full).
+        let seed = config
+            .seed
+            .unwrap_or(if consumer.is_pipeline_fault_tolerant() {
+                0
```

Comment:
Should the behavior of a connector depend on whether fault-tolerance is enabled?

If not, there could be for example: (a) make it `unwrap_or(0)` independent of fault-tolerance, or (b) make it situational by throwing an error if it is marked fault tolerant and it does not have a seed. (b) would make a connector not always fault-tolerant, I don't know if that fits in the current FT connector model though for them to be dynamically fault-tolerant.

For both (a) and the current implementation, the docs should be updated as well to reflect it as currently it states datagen uses a random seed unless one is provided: https://docs.feldera.com/connectors/sources/datagen

Reply:
I'll fix the behavior of the connector to match the documentation.

And I'll update the documentation to say that FT is now supported.